### PR TITLE
fix and improve signer-auth.md & reinitialization-attacks.md & pda-sharing.md &  owner-checks.md & closing-accounts.md & arbitrary-cpi.md

### PR DESCRIPTION
**Pull request by wuuer**
#### signer-auth.md

The way to get the seed "*ctx.bumps.get("vault").unwrap()" is outdated for the latest anchor version.
Unnecessary parameters in "accounts()" in the test typescript.

#### reinitialization-attacks.md
Magic number found on space calculation
“@project-serum/anchor” is outdated.
The two recommendedInitialization test cases do not use the same method to send transactions. 

#### pda-sharing.md
Magic number found on space calculation.
"token account" in the" Starter" section should refer to "associated token account"
"token account" in the "Add `initialize_pool_secure` instruction" section should refer to "associated token account"
Some typescript code is not synced to the project.

#### owner-checks.md
Magic number found on space calculation.
Unnecessary parameters are found in the test typescript.
Two Different ways to send transactions are found.

#### duplicate-mutable-accounts.md
Magic number found on space calculation.
"BorshDeserialize" and "BorshSerialize" are outdated.
Test descriptions are not clear enough.

#### closing-accounts.md
Use [InitSpace](https://docs.rs/anchor-lang/latest/anchor_lang/derive.InitSpace.html) to calculate space needed for accounts.
Delete "force_defund"  because “CLOSED_ACCOUNT_DISCRIMINATOR” was removed in the latest version of anchor-lang
Add the new secure instruction for account closing in the "Secure account closing" section .
Delete Unnecessary parameters found in the test typescript.
Change the Logic of account closing "Sets the account discriminator to the `CLOSED_ACCOUNT_DISCRIMINATOR` variant" to "Assigning the owner of the account to the System Program and rellocating the
 size of the account's data with 0 bytes."
#### arbitrary-cpi.md
The anchor cpi lesson link is invalid.
Magic number found on space calculation.

#### signer-auth.md

Update the latest way to get the seed.
replace "init" with "init_if_needed" for vault field in the struct InitializeVault<'info>.
delete unnecessary parameters "accounts()" in the test typescript.
Update the project repo links.

#### reinitialization-attacks.md
Use [InitSpace](https://docs.rs/anchor-lang/latest/anchor_lang/derive.InitSpace.html) to calculate space needed for accounts.
Use "@coral-xyz/anchor" instead of "@project-serum/anchor" for the test typescript.
Consistently use "rpc()" to send transactions in the test typescript.
Update BDD style for the test typescript.

**Also, I made a [PR](https://github.com/Unboxed-Software/solana-reinitialization-attacks/pull/2) for [solana-reinitialization-attacks](https://github.com/Unboxed-Software/solana-reinitialization-attacks/tree/starter) starter branch and a [PR](https://github.com/Unboxed-Software/solana-reinitialization-attacks/pull/3) for [solana-reinitialization-attacks](https://github.com/Unboxed-Software/solana-reinitialization-attacks/tree/solution) solution branch
which must be sync with this PR.**

#### pda-sharing.md
Use [InitSpace](https://docs.rs/anchor-lang/latest/anchor_lang/derive.InitSpace.html) to calculate space needed for accounts.
change "token account" in the "Starter" section to "associated token account"
change "token account" in the "Add `initialize_pool_secure" section to "associated token account"

**Also, I made a [PR](https://github.com/Unboxed-Software/solana-pda-sharing/pull/1) for [solana-pda-sharing](https://github.com/Unboxed-Software/solana-pda-sharing/tree/starter) starter branch and a [PR](https://github.com/Unboxed-Software/solana-pda-sharing/tree/solution) for [solana-pda-sharing](https://github.com/Unboxed-Software/solana-pda-sharing/pull/2) solution branch
which must be synced with this PR.**

#### owner-checks.md
Use [InitSpace](https://docs.rs/anchor-lang/latest/anchor_lang/derive.InitSpace.html) to calculate space needed for accounts.
Delete Unnecessary parameters found in the test typescript.
Consistently use "rpc()" as sending transactions in the test typescript.

**Also, I made a [PR](https://github.com/Unboxed-Software/solana-owner-checks/pull/1) for [solana-owner-checks](https://github.com/Unboxed-Software/solana-owner-checks/tree/starter) starter branch and a [PR](https://github.com/Unboxed-Software/solana-owner-checks/pull/2) for [solana-owner-checks](https://github.com/Unboxed-Software/solana-owner-checks/tree/solution) solution branch
which must be synced with this PR.**

#### duplicate-mutable-accounts.md
Use [InitSpace](https://docs.rs/anchor-lang/latest/anchor_lang/derive.InitSpace.html) to calculate space needed for accounts.
Replace "BorshDeserialize" and "BorshSerialize" with "AnchorDeserialize" and "AnchorSerialize".
Make test descriptions more clear.

**Also, I made a [PR](https://github.com/Unboxed-Software/solana-duplicate-mutable-accounts/pull/3) for [solana-duplicate-mutable-accounts](https://github.com/Unboxed-Software/solana-duplicate-mutable-accounts/tree/starter) starter branch and a [PR](https://github.com/Unboxed-Software/solana-duplicate-mutable-accounts/pull/4) for [solana-duplicate-mutable-accounts](https://github.com/Unboxed-Software/solana-duplicate-mutable-accounts/tree/solution) solution branch
which must be synced with this PR.**

#### closing-accounts.md
Use InitSpace to calculate space needed for accounts.
Delete "force_defund"  because “CLOSED_ACCOUNT_DISCRIMINATOR” was removed in the latest version of anchor-lang.
Add the new secure instruction for account closing in the "Secure account closing" section.
Change the Logic of closing account "Sets the account discriminator to the `CLOSED_ACCOUNT_DISCRIMINATOR` variant" to "Assigning the owner to the System Program".

**Also, I made a [PR](https://github.com/Unboxed-Software/solana-closing-accounts/pull/1) for [solana-closing-accounts](https://github.com/Unboxed-Software/solana-closing-accounts/tree/starter) starter branch and a [PR](https://github.com/Unboxed-Software/solana-closing-accounts/pull/2) for [solana-closing-accounts](https://github.com/Unboxed-Software/solana-closing-accounts/tree/solution) solution branch
which must be synced with this PR.**

#### arbitrary-cpi.md
Fix the anchor cpi lesson link.
Use [InitSpace](https://docs.rs/anchor-lang/latest/anchor_lang/derive.InitSpace.html) to calculate space needed for accounts.

**Also, I made a [PR](https://github.com/Unboxed-Software/solana-arbitrary-cpi/pull/1) for [solana-arbitrary-cpi](https://github.com/Unboxed-Software/solana-arbitrary-cpi/tree/starter) starter branch and a [PR](https://github.com/Unboxed-Software/solana-arbitrary-cpi/pull/2) for [solana-arbitrary-cpi](https://github.com/Unboxed-Software/solana-arbitrary-cpi/tree/solution) solution branch
which must be synced with this PR.**

### CODE REVIEW

**Comment by mikemaccana** [comment]
I've reviewed the PR on the other repo, but also forked it (and all branches to https://github.com/solana-developers/solana-signer-auth/ and invited you.

File: content/courses/program-security/signer-auth.md [hunk 1]
```diff
@@ -339,7 +339,9 @@ pub mod signer_authorization {
 #[derive(Accounts)]
 pub struct InitializeVault<'info> {
     #[account(
-        init,
+        // We use init_if_needed here for the test. Otherwise, the test will result in an error "Already in use" after testing once.
+        // Use "init" if you want to ensure that the "initialize_vault" function runs only once.
+        init_if_needed,
```
Comment (mikemaccana) [inline content/courses/program-security/signer-auth.md:344]: See comments in your PR to the other repo.

**Review by mikemaccana — CHANGES_REQUESTED** [review]
Great work. Nearly ready to go in. Just some minor bits.

File: content/courses/program-security/signer-auth.md (hunk 1 above)
Comment (wuuer) [reply content/courses/program-security/signer-auth.md:344]: see https://github.com/Unboxed-Software/solana-signer-auth/pull/3#discussion_r1735474018

**Comment by wuuer** [comment]
> I've reviewed the PR on the other repo, but also forked it (and all branches to https://github.com/solana-developers/solana-signer-auth/ and invited you.
@mikemaccana  I made PRs to this repo and changed the links in the course to refer to this repo.
Please do the same things:
1. Clone [solana-reinitialization-attacks](https://github.com/Unboxed-Software/solana-reinitialization-attacks/tree/starter) including all the branches.
2. Clone [solana-pda-sharing](https://github.com/Unboxed-Software/solana-pda-sharing/tree/starter) including all the branches.

**Comment by mikemaccana** [comment]
great work @wuuer!

1. Great work! Let's jump into telegram or X and DM for faster chat - I'm mikemaccana on either. 

2. You have the same code in this PR and #423 - which should I review? There's no need to open two PRs for the same work. 

3. New repos made:

https://github.com/solana-developers/reinitialization-attacks/
https://github.com/solana-developers/pda-sharing
https://github.com/solana-developers/closing-accounts

https://github.com/Unboxed-Software/solana-duplicate-mutable-accounts PRs need rebasing, as someone else already submitted updates. 

I could not do https://github.com/Unboxed-Software/solana-owner-checks, see https://github.com/Unboxed-Software/solana-owner-checks/pull/2

**Comment by wuuer** [comment]
> great work @wuuer!
> 
> 2. You have the same code in this PR and [fix and improve program-configuration.md #423](https://github.com/solana-foundation/developer-content/pull/423) - which should I review? There's no need to open two PRs for the same work.

@mikemaccana  
This PR(https://github.com/solana-foundation/developer-content/pull/372) is for program-security course content.
This PR(https://github.com/solana-foundation/developer-content/pull/423) is for program-optimization course content.
Do you agree with me doing this? If not, I will merge these two PRs.

**Review by mikemaccana — CHANGES_REQUESTED** [review]
@wuuer sorry about the wait here - we had a 24-person in-room class with new content in NY, then Rustconf then Breakpoint.

Re: your recent comment, some of your PRs are shortlisted for some lessons but yes, it's much easier to do one PR per lesson as others have done. There's no guarantee that someone that wins one lesson will win another one so we want to be able to merge and review them independently.

After you split them out, link them here with the lesson they are for.

**Comment by wuuer** [comment]
> Update the latest way to get the seed.
> replace "init" with "init_if_needed" for vault field in the struct InitializeVault<'info>.
> delete unnecessary parameters "accounts()" in the test typescript.
> Update the project repo links.

@mikemaccana

**Comment by wuuer** [comment]
> @wuuer sorry about the wait here - we had a 24-person in-room class with new content in NY, then Rustconf then Breakpoint.
> 
> Re: your recent comment, some of your PRs are shortlisted for some lessons but yes, it's much easier to do one PR per lesson as others have done. There's no guarantee that someone that wins one lesson will win another one so we want to be able to merge and review them independently.
> 
> After you split them out, link them here with the lesson they are for.

@mikemaccana 
[signer-auth](https://github.com/solana-foundation/developer-content/pull/526/)

File: content/courses/program-optimization/program-configuration.md [hunk 2]
```diff
@@ -1,13 +1,13 @@
 ---
-title: Program Configuration
+title: admin configuration
```
Comment (mikemaccana) [inline content/courses/program-optimization/program-configuration.md:2]:
```suggestion
title: Admin configuration
```

File: content/courses/program-optimization/program-configuration.md [hunk 3]
```diff
@@ -1,13 +1,13 @@
 ---
-title: Program Configuration
+title: Admin configuration
```
Comment (mikemaccana) [inline content/courses/program-optimization/program-configuration.md:2]: @wuuer this is great, sorry it's the last big PR and I've just gotten around to reviewing it.

File: content/courses/program-optimization/program-configuration.md [hunk 4]
```diff
@@ -866,51 +859,50 @@ pub fn payment_handler(ctx: Context<Payment>, amount: u64) -> Result<()> {
...
   assert.strictEqual(
     (
-      await program.account.programConfig.fetch(programConfig)
+      await program.account.adminConfig.fetch(adminConfig)
     ).feeBasisPoints.toNumber(),
     100,
   );
   assert.strictEqual(
-    (await program.account.programConfig.fetch(programConfig)).admin.toString(),
+    (await program.account.adminConfig.fetch(adminConfig)).admin.toString(),
```
Comment (mikemaccana) [inline content/courses/program-optimization/program-configuration.md:905]: For these strictEquals, it's a little busy, just add a variable name rather than the inline await.

```
const adminConfig = await program.account.adminConfig.fetch(adminConfig);
assert.strictEqual(adminConfig.admin, wallet.publicKey.toString())
```

File: content/courses/program-optimization/program-configuration.md [hunk 5]
```diff
@@ -919,10 +911,10 @@ it("Initialize Program Config Account", async () => {
 The second test verifies that the payment instruction is working correctly, with
 the fee being sent to the fee destination and the remaining balance being
 transferred to the receiver. Here we update the existing test to include the
-`programConfig` account.
+`adminConfig` account.
 
 ```typescript
-it("Payment completes successfully", async () => {
+it("Payment should complete successfully", async () => {
```
Comment (mikemaccana) [inline content/courses/program-optimization/program-configuration.md:917]: Grammar is off with these. 

Either:

> it("should complete successfully"

(if describe was 'payment')

or

> test("payment completes sucessfully")

See https://github.com/solana-foundation/developer-content/blob/main/CONTRIBUTING.md#jsts

Do this with all the tests.

File: content/courses/program-optimization/program-configuration.md [hunk 6]
```diff
@@ -955,67 +947,72 @@ it("Payment completes successfully", async () => {
...
-The fourth test tries to update the fee on the program config account, where the
-admin is not the one stored on the program config account, and this should fail.
+The fourth test tries to update the fee on the admin config account, where the
+admin is not the one stored on the admin config account, and this should fail.
 
 ```typescript
-it("Update Program Config Account with unauthorized admin (expect fail)", async () => {
+it("Admin Config Update with unauthorized admin should throw an exception", async () => {
   try {
     const tx = await program.methods
```
Comment (mikemaccana) [inline content/courses/program-optimization/program-configuration.md:978]: Use transaction, instruction, signature,  error, etc everywhere. See 'use full names' in https://github.com/solana-foundation/developer-content/blob/main/CONTRIBUTING.md#code

File: content/courses/program-security/pda-sharing.md [hunk 7]
```diff
@@ -231,13 +233,14 @@ program accounts.
 #### 1. Starter
 
 To get started, download the starter code on the `starter` branch of
-[this repository](https://github.com/Unboxed-Software/solana-pda-sharing/tree/starter).
+[this repository](https://github.com/solana-developers/pda-sharing/tree/starter).
```
Comment (mikemaccana) [inline content/courses/program-security/pda-sharing.md:236]:
```suggestion
[the `pda-sharing` repository](https://github.com/solana-developers/pda-sharing/tree/starter).
```

File: content/courses/program-optimization/program-configuration.md [hunk 8]
```diff
@@ -1027,52 +1024,6 @@ of [the same repository](https://github.com/Unboxed-Software/solana-admin-instr
 
 ## Challenge
 
-Now it's time for you to do some of this on your own. We mentioned being able to
```
Comment (mikemaccana) [inline content/courses/program-optimization/program-configuration.md:1030]: Let me know why you got rid of this!

File: content/courses/program-security/closing-accounts.md [hunk 9]
```diff
@@ -165,73 +145,6 @@ pub struct Data {
 }
 ```
 
-Note that the example above is using Anchor's `CLOSED_ACCOUNT_DISCRIMINATOR`.
```
Comment (mikemaccana) [inline content/courses/program-security/closing-accounts.md:168]: Likewise tell me about why you got rid of this section.

**Review by mikemaccana — CHANGES_REQUESTED** [review]
This is some solid work, if you can address the small issues below you'll win at least another one of the bounties.

**Comment by wuuer** [comment]
> This is some solid work, if you can address the small issues below you'll win at least another one of the bounties.

@mikemaccana . As [0xCipherCoder](https://github.com/solana-foundation/developer-content/pull/480) won the lesson `program-configuration`, I will not address these issues.

Please let me know if there are other PRs of mine left to win the Superteam earn 😉

File: content/courses/program-security/closing-accounts.md (hunk 9 above)
Comment (wuuer) [reply content/courses/program-security/closing-accounts.md:168]: “CLOSED_ACCOUNT_DISCRIMINATOR” was removed in the latest version of anchor-lang
https://github.com/coral-xyz/anchor/pull/2726#issue-2038294408

File: content/courses/program-security/pda-sharing.md (hunk 7 above)
Comment (wuuer) [reply content/courses/program-security/pda-sharing.md]: @mikemaccana Commit suggestion Accepted

File: content/courses/program-security/closing-accounts.md (hunk 9 above)
Comment (mikemaccana) [reply content/courses/program-security/closing-accounts.md:168]: Brilliant! You can resolve this (also if you see any other references to `CLOSED_ACCOUNT_DISCRIMINATOR`)

**Comment by mikemaccana** [comment]
@wuuer I sent you an email with details on your 3 official wins!

File: content/courses/program-optimization/program-configuration.md [hunk 10]
```diff
@@ -422,25 +419,25 @@ We'll quickly learn while testing our program that it could benefit from the
...
 
-### 1. Starter
+#### 1. Starter
 
-Download the starter code from
-the [`starter` branch of this repository](https://github.com/solana-developers/admin-instructions/tree/starter).
-The code contains a program with a single instruction handler and a single test
-in the `tests` directory.
+Download the starter code from the `starter` branch
+of [this repository](https://github.com/Unboxed-Software/solana-admin-instructions/tree/starter).
```
Comment (mikemaccana) [inline content/courses/program-optimization/program-configuration.md:425]: Why did you change this? Per the Superteam Earn instructions were are retiring the unboxed repos. That repo has moved, see the README on https://github.com/Unboxed-Software/solana-admin-instructions/

File: content/courses/program-optimization/program-configuration.md [hunk 11]
```diff
@@ -537,7 +527,7 @@ pub mod config {
 Next, add the `local-testing` feature to the `Cargo.toml` file located in
 `/programs`.
 
-```shell
+```
```
Comment (mikemaccana) [inline content/courses/program-optimization/program-configuration.md:530]:
```suggestion
```toml
```

File: content/courses/program-optimization/program-configuration.md [hunk 12]
```diff
@@ -946,25 +930,24 @@ it("Payment should complete successfully", async () => {
...
-      sender,
-    ]);
+  await anchor.web3.sendAndConfirmTransaction(connection, tx, [sender]);
 
-    const senderBalance = await getAccount(connection, senderTokenAccount);
-    const feeDestinationBalance = await getAccount(connection, feeDestination);
-    const receiverBalance = await getAccount(connection, receiverTokenAccount);
+  assert.strictEqual(
+    (await connection.getTokenAccountBalance(senderTokenAccount)).value
+      .uiAmount,
```
Comment (mikemaccana) [inline content/courses/program-optimization/program-configuration.md:937]: Just use a variable rather than inline the await, it's easier to read.

File: content/courses/program-optimization/program-configuration.md [hunk 13]
```diff
@@ -946,25 +930,24 @@ it("Payment should complete successfully", async () => {
...
-    );
-    expect(Number(receiverBalance.amount)).to.equal(
-      (PAYMENT_AMOUNT * (10000 - INITIAL_FEE_BASIS_POINTS)) / 10000,
-    );
-  } catch (error) {
-    console.error("Payment failed:", error);
-    throw error;
-  }
+  assert.strictEqual(
+    (await connection.getTokenAccountBalance(feeDestination)).value.uiAmount,
```
Comment (mikemaccana) [inline content/courses/program-optimization/program-configuration.md:942]: Same here (and anywhere else with inlined awaits)

File: content/courses/program-optimization/program-configuration.md [hunk 14]
```diff
@@ -1035,18 +1018,17 @@ Program log: Right:
...
-
-  4 passing (22s)
+4 passing (8s)
 ```
 
 And that's it! You've made the program a lot easier to work with moving forward.
 If you want to take a look at the final solution code you can find it on
-the [`solution` branch of the same](https://github.com/solana-developers/admin-instructions/tree/solution).
+the `solution` branch
+of [the same repository](https://github.com/Unboxed-Software/solana-admin-instructions/tree/solution).
```
Comment (mikemaccana) [inline content/courses/program-optimization/program-configuration.md:1027]: Use the right repo as mentioned above.

**Review by mikemaccana — CHANGES_REQUESTED** [review]
Just a reminder this has won program-security/closing-accounts.md, you can probably remove the changes to other lessons (maybe submit them later) and focus on program-security/closing-accounts.md

File: content/courses/program-optimization/program-configuration.md (hunk 10 above)
Comment (mikemaccana) [reply content/courses/program-optimization/program-configuration.md:425]: Actually ignore that it was for the wrong lesson.

File: content/courses/program-optimization/program-configuration.md (hunk 11 above)
Comment (mikemaccana) [reply content/courses/program-optimization/program-configuration.md:530]: Actually ignore that it was for the wrong lesson.

File: content/courses/program-optimization/program-configuration.md (hunk 12 above)
Comment (mikemaccana) [reply content/courses/program-optimization/program-configuration.md:937]: Actually ignore that it was for the wrong lesson.

File: content/courses/program-optimization/program-configuration.md (hunk 13 above)
Comment (mikemaccana) [reply content/courses/program-optimization/program-configuration.md:942]: Actually ignore that it was for the wrong lesson.

File: content/courses/program-optimization/program-configuration.md (hunk 14 above)
Comment (mikemaccana) [reply content/courses/program-optimization/program-configuration.md:1027]: Actually ignore that it was for the wrong lesson.
